### ISSUE TITLE
A non-REGISTERing PBX origination within 30 seconds of a previous origination results in a 403 response vs. 407

### ISSUE THREAD

**Issue by plfrancis** [Metaswitch/sprout#1217]
If a non-REGISTERing PBX subscriber originates a call, abandons (e.g. due to a misdialed number), then re-originates within 30 seconds, a 403 response will be returned instead of the expected 407. This will lead to an unexpected call failure, vs. a "silent" challenge/re-authorization and successful re-origination.

**Comment by AlexHockey** [comment]
I'm unassigning myself as I only have bandwidth to work on critical bugs ATM.

**Comment by rkd-msw** [comment]
Can you provide diagnostics?

**Comment by rkd-msw** [comment]
(I'm thinking SAS trace for the two calls, plus a diags bundle if possible.)

**Comment by plfrancis** [comment]
Requested diagnostics have been uploaded.

**Comment by rkd-msw** [comment]
I've taken a look. In the failed trace:

* there's an INVITE, then a 407 challenge, then an authenticated INVITE with the header `Proxy-Authorization: Digest username="9165191900@vm1.richlab.datcon.co.uk",realm="vm1.richlab.datcon.co.uk",cnonce="6b8b4567",nc=00000001,qop=auth,uri="sip:192.168.140.150:6060",nonce="6e7b886a610608c0",response="38aa8d5670a7d8e9aa9dc89a17979eef",algorithm=MD5,opaque="280843ce4bc1a770"`
* that call then ends, and a new call starts, but this time it starts using an authenticated INVITE which reuses the original nonce, sets a new client nonce and increments the nonce count, with the header `Proxy-Authorization: Digest username="9165191900@vm1.richlab.datcon.co.uk",realm="vm1.richlab.datcon.co.uk",cnonce="327b23c6",nc=00000002,qop=auth,uri="sip:192.168.140.150:6060",nonce="6e7b886a610608c0",response="22513055b65799c32e445a085bf545f6",algorithm=MD5,opaque="280843ce4bc1a770"`.

This second header has an incorrect response field (as far as I can tell), which is causing the Forbidden response. I wrote a script (attached) to check the response values - it generates 38aa8d5670a7d8e9aa9dc89a17979eef for the first one, which matches, but 4f42ec8656654247ec2d77fe551f96e0 for the second one.

In the working traces, because of the 30 second delay, our stored nonce has timed out, so we generate a new nonce and a 407 challenge, and the client generates the right response - possibly that indicates a client bug with nonce counts greater than 1?

More generally, though, I don't think we support nonce reuse and nonce counts at the moment (there's a backlog story for it), so can you retry this test with the client set up to not reuse nonces?
[checkauth.py.txt](https://github.com/Metaswitch/sprout/files/11424/checkauth.py.txt)

**Comment by AlexHockey** [comment]
It might not be possible to turn off nonce-count incrementing in the client. Can we find one that calculates the response correctly and test with that?

**Comment by plfrancis** [comment]
I modified SIPp to support this test case. Unfortunately I did not notice that it was pulling the method for the auth. calculation from the last message received vs. the current message (or the last received message with an auth="true" tag). I've fixed this issue and now the INVITE with "pre-calculated" credentials is accepted without a challenge. The only thing remaining here is to update section 5.1.1 of the specification to indicate that requests with "pre-calculated" credentials will be accepted up to the auth. time-out. Passing to Alex for that bit and to close out the issue. Sorry for the confusion.
